openTSNE/nearest_neighbors.py

```diff
@@ -470,8 +476,14 @@ def build(self):
         timer.__enter__()
 
         # These values were taken from UMAP, which we assume to be sensible defaults
-        n_trees = 5 + int(round((data.shape[0]) ** 0.5 / 20))
-        n_iters = max(5, int(round(np.log2(data.shape[0]))))
+        if self.n_trees is None:
```

Comment:
So this bit of code here looks like it would only support three parameters for NNDescent: `n_trees`, `n_iters`, and `max_candidates`. pynndescent supports a whole range of other parameters as well:
https://github.com/lmcinnes/pynndescent/blob/master/pynndescent/pynndescent_.py#L673-L697

Ideally, I'd be able to specify any of these arguments. Instead of storing the nndescent-specific parameters as class attributes (`self.n_trees`, `self.n_iters`), I think it would be better to store a dictionary of `self.knn_kwargs`, and use that. In this case, in case a KNN library deprecates its parameters and stops using it, we won't have to go through a deprecation cycle as well. And this would make specifying any potential newly introduced arguments as simple as just passing it into the dictionary without having to update openTSNE. I like this approach better since openTSNE remains mostly agnostic to the KNN libraries, which can update their parameters independently of openTSNE.

So maybe running something like `NNDescent(data, k=15, knn_kwargs={"n_trees": 50})` instead of the way you have it now: NNDescent(data, k=15, n_trees=50).

As a side benefit, this would also make it explicit to users that this is a dict that gets passed through to the underlying implementation, and that we don't handle the parameters in any special way. Similarly to the logic of `metric_params`.

Reply:
Okay, that sounds like a good idea to me.  I took that approach so that we can still use the `**kwargs` to pass parameters to the super class.  I like the approach with preserving the dict instead and only splatting it later down the line for all classes.  

Should the default parameters be preserved?  As in, should we check if they are present in the dict and if not take the defaults that are currently used?  I personally would like to use the default of pynndescent, but that would potentially break backwards compatibility, so I would tend to do the check and by default initialize with the current defaults still.